Silicon/NXP/iMXPlatformPkg/Drivers/SmbiosConfigDxe/SmbiosConfigDxe.c

```diff
@@ -0,0 +1,878 @@
+  //
+  // Find Smbios.ini file starting with Pcd device path
+  //
+  DevicePathText = (CONST CHAR16 *) FixedPcdGetPtr (PcdSmbiosOverrideDevicePath);
+  if ((DevicePathText == NULL) || (*DevicePathText == L'\0')) {
+    Status = EFI_INVALID_PARAMETER;
+    DEBUG ((DEBUG_ERROR, "%a: PcdStorageMediaPartition is unspecified\n", __FUNCTION__));
+    goto Exit;
+  }
+
+  DevicePath = ConvertTextToDevicePath (DevicePathText);
+  if (!DevicePath) {
+    DEBUG ((DEBUG_ERROR, "%a: Could not convert DevicePathText\n", __FUNCTION__));
+    return FALSE;
+  }
+
+  IntermediateDevicePath = DevicePath;
+  Status = gBS->LocateDevicePath (
+    &gEfiSimpleFileSystemProtocolGuid,
+    &IntermediateDevicePath,
+    &MediaHandle);
+
+  if (Status == EFI_NOT_FOUND) {
+    DEBUG ((DEBUG_INFO,"%a: %s FAT partition is not ready yet\n", __FUNCTION__, DevicePathText));
+    goto Exit;
+  }
+
+  if (EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_ERROR, "%a: gBS->LocateDevicePath() failed. (Status=%r)\n", __FUNCTION__, Status));
+    goto Exit;
+  }
+
+  Status = gBS->HandleProtocol (
+    MediaHandle,
+    &gEfiSimpleFileSystemProtocolGuid,
+    (VOID**)&Fs
+    );
+
+  if (EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_ERROR, "%a: gBS->HandleProtocol() failed. (Status=%r)\n", __FUNCTION__, Status));
+    goto Exit;
+  }
+
+  Status = Fs->OpenVolume (Fs, &RootVolume);
+  if (EFI_ERROR(Status)) {
+    DEBUG ((DEBUG_ERROR, "%a: Fs->OpenVolume() failed. (Status=%r)\n", __FUNCTION__, Status));
+    goto Exit;
+  }
+
+  Status = RootVolume->Open (
+    RootVolume,
+    &File,
+    L"Smbios.ini",
+    EFI_FILE_MODE_READ,
+    0);
+
+  if (EFI_ERROR(Status)) {
+    DEBUG ((DEBUG_ERROR, "%a: RootVolume->Open() failed. (Status=%r)\n", __FUNCTION__, Status));
+    goto Exit;
+  }
+
+  //
+  // Read File Info
+  // 
+  FileInfoSize = sizeof (EFI_FILE_INFO) + 1024;
+  FileInfo = AllocateZeroPool (FileInfoSize);
+  if (FileInfo == NULL) {
+    DEBUG ((DEBUG_ERROR, "%a: FileInfo allocation failed\n", __FUNCTION__));
+    goto Exit;
+  }
+
+  Status = File->GetInfo (
+                  File,
+                  &gEfiFileInfoGuid,
+                  &FileInfoSize,
+                  FileInfo
+                  );
+  if (EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_ERROR, "%a: Failed to get File Info (Status=%r)\n", __FUNCTION__, Status));
+    goto Exit;
+  }
+
+  BufferSize = (UINTN) FileInfo->FileSize + sizeof (CHAR16);
+  Buffer = AllocateZeroPool (BufferSize);
+  if (Buffer == NULL) {
+    DEBUG ((DEBUG_ERROR, "%a: Buffer allocation failed \n", __FUNCTION__));
+    goto Exit;
+  }
+
+  Status = File->Read (
+                File,
+                &BufferSize,
+                Buffer
+                );
+  if (EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_ERROR, "%a: File Read failed (Status=%r)\n", __FUNCTION__, Status));
+    goto Exit;
+  }
+
+  //
+  // Scan through the INI file and add its contents into the SmbiosOverride linked list
+  //
+  Status = PopulateSmbiosOverrideList (Buffer, BufferSize);
```

Comment:
Add a function to validate the INI syntax itself before populating the override list.
Otherwise you end up with a partially filled override list.